pkg/crds/crds.go

```diff
@@ -0,0 +1,205 @@
+// Package crds is used for installing rancher CRDs.
+package crds
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/rancher/rancher/pkg/fleet"
+	"github.com/rancher/wrangler/pkg/crd"
+	"github.com/rancher/wrangler/pkg/yaml"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// yamlDir directory holding CRDs yaml files.
+	yamlDir = "yamls"
+
+	// readyDuration time to wait for CRDs to be ready.
+	readyDuration = time.Minute * 1
+
+	k8sManagedByKey = "app.kubernetes.io/managed-by"
+	managerValue    = "rancher"
+)
+
+//go:embed yamls/*
+var crdFS embed.FS
+
+// EnsureRequired will ensure all required CRDs needed by Rancher based on the currently enabled features are installed and up to date.
+func EnsureRequired(ctx context.Context, cfg *rest.Config) error {
```

Comment:
Nit: It looks like you really only need the CustomResourceDefinitionInterface (`clientSet.ApiextensionsV1().CustomResourceDefinitions()`) in this function and the sub-functions that this calls. I would recommend accepting that interface as an arg rather than the struct for the following reasons: to follow the whole "accept interfaces return structs" standard, to allow for easier testing, and to not re-create the entire clientset here.

Reply:
As far as the testing goes, hard to argue with what you have written so far.

As far as the other parts go, I think it's tough to argue with the specifics of the current situation, meaning that this code only needs one particular resource (the CRD interface), and that particular resource is an interface. I think that it makes sense to use that fact to our advantage here, and not require that the consuming function produce a restConfig (particularly in the context of rancher, since early on in the boot process we extract these types of things and attach them to the wrangler context).